### PR TITLE
Bookmarks reorder

### PR DESCRIPTION
**Pull request by gusaiani**
### How to Test

1. Add bookmarks.
2. On the left sidebar, drag them up and down to reorder.
3. Reload the page.

Bookmark order should be preserved.

### CODE REVIEW

**Comment by mazameli** [comment]
@gusaiani I'm seeing some funny behavior with the dragging: https://www.loom.com/share/fd5348098e654100bbc85954cb8ef9d2

**Comment by gusaiani** [comment]
@mazameli, thanks very much.

*tldr:* I did something that seems to make things significantly better.

2 technical details worth articulating aloud:

1. Managing potential conflicts between dragging the item and clicking inside the item

To remove a bookmark, or to navigate to the bookmark's page, we have to just click inside the item.
This can be interpreted as an intention to drag the item.

What our library offers is a kind of timeout: right now it's set to 200 milliseconds.
So if you want to just click and are not too slow to do it, you're good.
And if you want to drag, merely the time holding the mouse down activates the drag.

I've played around with the milliseconds, and with 200 it works pretty nicely now in my machine, though not absolutely perfectly. I can sometimes intentionally make it behave as it has for you.

2. The drag-and-drop library itself is deprecated

[This is the one](https://github.com/clauderic/react-sortable-hoc).

They recommend moving to [https://github.com/clauderic/dnd-kit](dnd-kit).

I've checked out dnd-kit a little. 
Seems somewhat different from our current one, enough that making the change should be done as a mini project, and maybe one-by-one in the drag-drops we have.

At the same time, we already use a second dnd library in our repo.
It's a more complicated one.

I think the takeaway from all the above is:
if what we have now, or with tweaks to the `pressDelay` timeout, does the trick, let's keep it as is for this PR.

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/sortable.css [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.react-sortable-hoc-helper.sorting {
```
Comment (ranquild) [inline frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/sortable.css:1]: Could we move this to emotion? E.g. 

const Component = styled.div```
  .sorting {
    
  }
```;

File: frontend/src/metabase/entities/bookmarks.js [hunk 2]
```diff
@@ -23,7 +23,18 @@ const Bookmarks = createEntity({
   objectSelectors: {
     getIcon,
   },
-
+  objectActions: {
```
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/entities/bookmarks.js]: @ranquild 

Thank you.

Updated and I've added a couple of lines of comments in `lib/entities` with what you wrote above.
Maybe it can be improved, but I think it's better to have something than nothing.

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.styled.tsx [hunk 3]
```diff
@@ -25,5 +44,25 @@ export const SidebarBookmarkItem = styled(SidebarLink)`
...
+    }
   }
+
+  ${props =>
+    props.isSorting &&
+    css`
+      &:hover {
+        background: white;
+
+        svg {
```
Comment (ranquild) [inline frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.styled.tsx:59]: You can use 

```
${DragIcon} {
  color: ...
}
```;

to style the icon instead of relying on html tags

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/sortable.css (hunk 1 above)
Comment (gusaiani) [reply frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/sortable.css:1]: @ranquild It may be possible, but I think it would be a lot of work for little gain, and maybe not possible.

This class is added by `react-sortable-hoc` to the cloned dragged component, when we are dragging.

And it's not something, afaik, we can write a custom component and inject styled-components/emotion.

Also this library has been deprecated for a while now, so even less incentive to invest in this I suppose.

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.tsx [hunk 4]
```diff
@@ -28,56 +35,102 @@ interface CollectionSidebarBookmarksProps {
...
 }: CollectionSidebarBookmarksProps) => {
+  const [orderedBookmarks, setOrderedBookmarks] = useState(bookmarks);
+  const [isSorting, setIsSorting] = useState(false);
+
+  useEffect(() => {
+    setOrderedBookmarks(bookmarks);
+  }, [bookmarks]);
+
   const onToggleBookmarks = useCallback(isVisible => {
     localStorage.setItem("shouldDisplayBookmarks", String(isVisible));
```
Comment (ranquild) [inline frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.tsx:108]: It would be great to move it to user-settings on the backend

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/sortable.css (hunk 1 above)
Comment (ranquild) [reply frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/sortable.css:1]: I don't think you would need to change anything in how `sortable` works. Just use this `sorting` class in your `emotion` styles:

```
export const SidebarBookmarkItem = styled(...)`
  &.sorting {
    background-color: ${color("white")};
    border-bottom: 2px solid ${color("brand")};
  }
`;
```

**Review by ranquild — APPROVED** [review]
Seems to work correctly. It would be great to:
- Use component references instead of html tags in styles, e.g. `${DragIcon}`
- Remove the CSS file and move these styles to styled.tsx.

Could be also a follow-up PR.

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.styled.tsx (hunk 3 above)
Comment (gusaiani) [reply frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.styled.tsx:59]: Thank you @ranquild.

I've just tried this from your feedback.

Here we're controlling the style both for the DragIcon svg and the icon next to the name of the item.

I don't know if it's advantageous to refer to the components in this case, as it can get more verbose.
Since we are indeed styling the svg's, I'd believe that this way is terse and understandable.

Wdyt?

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.tsx (hunk 4 above)
Comment (gusaiani) [reply frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.tsx:108]: Yup, haven't forgotten that, thanks for the reminder.

**Comment by kulyk** [comment]
@gusaiani FYI, you can also use `distance` prop apart from `pressDelay` to distinguish click and drag events, maybe it'd work better, but I'm not sure. You can see more [here](https://github.com/clauderic/react-sortable-hoc#click-events-being-swallowed)

File: frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.styled.tsx (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase/nav/containers/MainNavbar/BookmarkList/BookmarkList.styled.tsx:59]: I don't have good arguments to justify using component references here. I just thought that it's not obvious that components could be used like that, and wanted to share this info.

**Comment by gusaiani** [comment]
> you can also use distance prop apart from pressDelay to distinguish click and drag events, maybe it'd work better, but I'm not sure. You can see more [here](https://github.com/clauderic/react-sortable-hoc#click-events-being-swallowed)

@kulyk oh thanks very much. I overlooked that one. Playing with it.

**Comment by mazameli** [comment]
@gusaiani Dragging looks nice now! I found one glitch where the order seems to not persist, but then on a hard refresh it appears to reflect the new order correctly. Examples here: https://www.loom.com/share/d0063ddd7726408690864d3b2d67dab8

**Review by mazameli — CHANGES_REQUESTED** [review]
Looks great except for the glitch I reported here: https://github.com/metabase/metabase/pull/21536#issuecomment-1099368420

**Comment by mazameli** [comment]
Looking now…

**Review by mazameli — APPROVED** [review]
Looking good!

**Comment by gusaiani** [comment]
@mazameli Thank you, thank you very much for catching that glitch I totally missed xD

**Comment by mazameli** [comment]
Stoked this got in! Nice work.
